Initial fragmentation

TimEvens opened the pull request:
Fragmentation support added.  Several files are also updated due to ```make format```

TimEvens committed:
Initial fragmentation updates, inculding make format

TimEvens committed:
Update build-tests to install pkg-config

TimEvens committed:
Fix build test issues

TimEvens committed:
header file change

TimEvens committed:
More include updates

TimEvens committed:
Revert header removal

GhostofCookie commented on include/quicr/hex_endec.h:
```diff
@@ -36,8 +36,7 @@ class HexEndec
   template<typename... UInt_ts>
```
Are these just formatting changes in this file?

TimEvens replied:
Yes, make format is becoming a problem.  I'm going to stop using it for now.

GhostofCookie commented on cmd/really/testLogger.h:
```diff
@@ -0,0 +1,11 @@
+#pragma once
+#include <mutex>
+#include <transport/logger.h>
+
+class testLogger : public qtransport::LogHandler {
```
We actually happen to have a logger interface in cantina, should we be using that instead of rolling something new? I know qmedia has something hand rolled as well.

TimEvens replied:
sure, let's do that in a different PR though as others should be changed to use a common one.

TimEvens committed:
Add log

GhostofCookie commented on src/quicr_client.cpp:
```diff
@@ -248,21 +434,30 @@ QuicRClient::handle(messages::MessageBuffer&& msg)
+            false,
+            std::move(datagram.media_data));
+        } else {
+          std::cout << "Name:" << datagram.header.name.to_hex()
+                    << ", not in namespace " << entry.first.to_hex()
+                    << std::endl;
```
Is this a helpful output? Shouldn't we output only if we never found a valid namespace?

TimEvens replied:
The else is called if it's not found, not valid.  This PR does add logger so that we can pass it down, but it doesn't update all the ```std::cout <<``` log messages. All of these should be cleaned up.

GhostofCookie commented on src/quicr_client.cpp:
```diff
@@ -227,6 +326,93 @@ QuicRClient::publishNamedObjectFragment(const quicr::Name& /* quicr_name */,
+  }
+
+  // Move to next buffer if reached max
+  if ((int)fragments[cindex].size() >= MAX_FRAGMENT_NAMES_PENDING_PER_BUFFER) {
```
Why c-style cast to int? Why isn't `MAX_FRAGMENT_NAMES_PENDING_PER_BUFFER` unsigned?

TimEvens replied:
No, its signed... although, it can be changed to unsigned. While the value will never be negative, it's a config parameter that should be validated upon input/config.

GhostofCookie replied:
Validation sounds good. I commented because it's generally [frowned upon to use c-style casting](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es49-if-you-must-use-a-cast-use-a-named-cast) these days (and in fact [avoiding cast when you can](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es48-avoid-casts)). Changing `MAX_FRAGMENT_NAMES_PENDING_PER_BUFFER` to unsigned should do that. And doing some validation on input/config as you suggest is good too.

TimEvens replied:
I just changed it to unsigned as it doesn't have to be signed. Cast removed.

GhostofCookie commented on src/quicr_client.cpp:
```diff
@@ -227,6 +326,93 @@ QuicRClient::publishNamedObjectFragment(const quicr::Name& /* quicr_name */,
+void
+QuicRClient::handle_pub_fragment(messages::PublishDatagram&& datagram)
+{
+  static int cindex = 1;
```
Should this be static? Are we guaranteed that by the end of the function cindex will be set to 1 again? Or do we not care?

TimEvens replied:
Yes, it's persistent and is maintained within this method.  It'll wrap back to 1 when reaching MAX FRAGMENT BUFFERS.  I suppose we can move it to the class var, although this is the only method that uses it.

GhostofCookie reviewed (commented):
I don't see anything too concerning in here, though it's hard to read with all the formatting changes on top. I'll hold off approval for now. Looks good!

My comments are optional, mostly just dumb questions.

paulej commented on cmd/really/CMakeLists.txt:
```diff
@@ -11,4 +12,4 @@ set_target_properties(really
         CXX_STANDARD_REQUIRED YES
-        CXX_EXTENSIONS ON)
\ No newline at end of file
+        CXX_EXTENSIONS ON)
```
For portability, we should not turn this on.  While portability is not critical now, it might be later.  What extensions are we using we cannot live without?

TimEvens replied:
Can turn it off, but I believe it's enabled by default.  If we remove it completely, wouldn't it be ON?

TimEvens replied:
I'll leave this as is for now since it's everywhere.  Let's clean this up if needed in all the CMakeLists.txt files.

paulej commented on cmd/really/testLogger.cpp:
```diff
@@ -0,0 +1,35 @@
+void testLogger::log(qtransport::LogLevel level, const std::string &string) {
+  const char *lvl;
+  switch (level) {
+  case qtransport::LogLevel::fatal:
```
Just for consistency with syslog, I called this "critical" in the Cantina logger.  I'd suggest the same here for that same reason.

TimEvens replied:
We can refactor that to match, but that will need to first be done under transport. Let's do that in a different PR considering that may involve several parent modules that use transport.

TimEvens committed:
Remove debug log message

TimEvens committed:
Fix last fragment when remaining is zero

TimEvens committed:
Remove c-style cast

TimEvens committed:
Fix on publish delegate logging

With multiple subscriptions, a log message will be logged for
each subscription namespace that doesn't match. This
disables that log for now.

TimEvens committed:
Updte to latest transport as well

GhostofCookie commented on src/quicr_server.cpp:
```diff
@@ -140,47 +130,23 @@ QuicRServer::subscriptionEnded(const quicr::Namespace& /* quicr_namespace */,
-                             [[maybe_unused]] uint64_t best_before,
                              [[maybe_unused]] bool use_reliable_transport,
-                             bytes&& data)
+                             const messages::PublishDatagram& datagram)
```
Should this be messages::PublishDatagram&&?

TimEvens replied:
No, it's not a move. It's retained.

GhostofCookie replied:
We don't want to move the datagram? Not even to reduce copying the payload vector?

GhostofCookie commented on src/quicr_server.cpp:
```diff
@@ -226,15 +192,17 @@ QuicRServer::handle_publish(
-                             0,
-                             false,
-                             std::move(datagram.media_data));
+  //  std::ostringstream log_msg;
+  //  log_msg << "handle_publish: cid: " << context_id << " sid: " << mStreamId
+  //          << " gid: " << from_varint(datagram.header.group_id)
+  //          << " oid: " << from_varint(datagram.header.object_id)
+  //          << " off: " << from_varint(datagram.header.offset_and_fin);
+  //  log_handler.log(qtransport::LogLevel::info, log_msg.str());
```
We probably shouldn't push dead code. If we want it back we can get it from version control.

TimEvens replied:
It was there in case we need to add it back for debugging, but IMO I don't need it anymore, so I'll remove.

TimEvens committed:
Remove unused logging statement

shenning00 reviewed (approved):
That is a lot of changes.

Ran new-qmedia test code again `really`.

Then ran WxQ against `cloud` relay.
